### ISSUE TITLE
[HOLD for payment 2022-12-20] [HOLD for payment 2022-12-15] Request money - The list of currency is empty until you enter any letter in search field

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#13286]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to URL https://staging.new.expensify.com/
2. Login with any account
3. Click on FAB button -> Request money
4. Click on the currency icon

## Expected Result:
The currency list must be filled with a list of currencies

## Actual Result:
The list of currency is empty until you enter any letter in search field

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.36-0
**Reproducible in staging?:** y
**Reproducible in production?:** n
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/205375413-8cb694a7-82c5-4dcf-98d1-7c61f25153fc.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** Applause internal team
**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01cfd75a5dd3d584dd</li>
        <li>Upwork Job ID: 1606450618598191104</li>
        <li>Last Price Increase: 2022-12-24</li>
    </ul>
</details>

**Comment by s77rt** [comment]
### Proposal
```diff
diff --git a/src/libs/OptionsListUtils.js b/src/libs/OptionsListUtils.js
index bc9a4facd5..62fda478db 100644
--- a/src/libs/OptionsListUtils.js
+++ b/src/libs/OptionsListUtils.js
@@ -787,6 +787,10 @@ function getHeaderMessage(hasSelectableOptions, hasUserToInvite, searchValue, ma
  * @returns {Array}
  */
 function getCurrencyListForSections(currencyOptions, searchValue) {
+    if (!searchValue) {
+        return {currencyOptions};
+    }
+
     const filteredOptions = _.filter(currencyOptions, currencyOption => (
         isSearchStringMatch(searchValue, currencyOption.text)));
 ```

### Details
If the `searchValue` is empty return all the currency options

@0xmiroslav I don't think this is a regression, it's more like a hidden bug that has been revealed. If the search value is empty we should return nothing instead of everything unless explicitly specified. Also this will avoid the unnecessary `_filter` call and squeeze more performance.

**Comment by tgolen** [comment]
The proposal looks good to me 👍 I'm a bit conflicted on it being a regression or not.

I think it's technically a regression if it worked properly before and now it doesn't. I think this is just the nature of some regressions and it's why we have the checklist item:

>  I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like Avatar, I verified the components using Avatar are working as expected)

Since `isSearchStringMatch` was modified, then all the references to it should have been tested and this issue could have been caught.

**Comment by tgolen** [comment]
cc @cead22 for your thoughts on it since you reviewed the original PR

**Comment by s77rt** [comment]
@0xmiroslav Yes I did, I agree with you, I'm just saying that we have found a hidden bug and posted a solution above.

**Comment by situchan** [comment]
### **Proposal:**

https://github.com/Expensify/App/blob/main/src/libs/OptionsListUtils.js#L391-L401
```diff
function isSearchStringMatch(searchValue, searchText, participantNames = new Set(), isChatRoom = false) {
+   if (!searchValue.trim()) {
+        return true;
+   }
    const searchWords = _.compact(uniqFast([
        searchValue,
        ..._.map(searchValue.replace(/,/g, ' ').split(' '), word => word.trim()),
    ]));
    const valueToSearch = searchText && searchText.replace(new RegExp(/&nbsp;/g), '');
    return _.some(searchWords, (word) => {
        const matchRegex = new RegExp(Str.escapeForRegExp(word), 'i');
        return matchRegex.test(valueToSearch) || (!isChatRoom && participantNames.has(word));
    });
}
```
This allows no filter with endless blank text like all other popular search engines do

**Comment by s77rt** [comment]
@situchan This will still call `_filter`.
Also this will return everything if the search value is empty which may break other components.

> If the search value is empty we should return nothing instead of everything unless explicitly specified.

**Comment by parasharrajat** [comment]
@s77rt As this is caused due to https://github.com/Expensify/App/pull/13186 and the regression period for that PR is still active, the author will have to send a fix for it. 


We will revert the PR as it is blocking the app deployments. https://github.com/Expensify/App/issues/13286#issuecomment-1335768149



cc: @rushatgabhane You are C+ on that.

**Comment by s77rt** [comment]
@parasharrajat Shall I make another PR with the above fix then?

**Comment by parasharrajat** [comment]
Yeah, go ahead. But I haven't checked the proposal. I will go with the approval from @tgolen. 

@rushatgabhane will help further in testing and reviewing.

**Comment by s77rt** [comment]
@parasharrajat https://github.com/Expensify/App/pull/13295

**Comment by parasharrajat** [comment]
Please add this issue to that PR as well. 
Add a new line below the existing issue. 

```
$ https://github.com/Expensify/App/pull/13295
```

**Comment by s77rt** [comment]
Updated

**Comment by cead22** [comment]
> cc @cead22 for your thoughts on it since you reviewed the original PR

I agree with @tgolen 's take in this [comment](https://github.com/Expensify/App/issues/13286#issuecomment-1335847206). I'll keep an eye out for this type of testing in future reviews. In this case, should we expect screenshots of these other parts of the code in the PR, and really everything that was tested? I think we should

**Comment by tgolen** [comment]
Looks like the proposal is 🟢 so we can move forward with getting this fixed with a PR.

**Comment by cead22** [comment]
@s77rt @tgolen @parasharrajat I'm reviewing the PR now, but given this is only called from two places in the same file:
- Instead of checking if searchValue is empty and doing an early return, wouldn't it be better to do that check before calling getCurrencyListForSections? 
    - Here https://github.com/Expensify/App/blob/7cd32889a/src/pages/iou/IOUCurrencySelection.js#L44 we just need to update 
        - `const {currencyOptions} = OptionsListUtils.getCurrencyListForSections(this.getCurrencyOptions(this.props.currencyList), '');` with something like
        -  `const {currencyOptions} = this.getCurrencyOptions(this.props.currencyList);`
        - IMO this is better than passing an empty string to check for empty string and do an early return
    - Here https://github.com/Expensify/App/blob/7cd32889a/src/pages/iou/IOUCurrencySelection.js#L93 we have to update 
        - `const {currencyOptions} = OptionsListUtils.getCurrencyListForSections(this.getCurrencyOptions(this.props.currencyList), searchValue);` with something like
        - `const {currencyOptions} = searchValue ?OptionsListUtils.getCurrencyListForSections(this.getCurrencyOptions(this.props.currencyList)) : this.getCurrencyOptions(this.props.currencyList);`
- Instead of having this functionality that is only used in `IOUCurrencySelection.js` defined in `OptionsListUtils.js`, should we move it to `IOUCurrencySelection.js`?

**Comment by parasharrajat** [comment]
cc: @rushatgabhane :arrow_up:. I am not working on this.

**Comment by tgolen** [comment]
@cead22 I like your cleanup proposal, yeah. I agree that if the logic for the shared library is only used by one component, we should move that logic to the component itself.

**Comment by s77rt** [comment]
@cead22 Should I make this change in my PR?

**Comment by s77rt** [comment]
I don't think we should move `getCurrencyOptions` as it relies on `isSearchStringMatch`. It's seems better to keep it as it is
I will apply the changes suggested by @cead22 but i will keep my check on https://github.com/Expensify/App/issues/13286#issuecomment-1335826679 in case this function was called by another compoenent in the future
let me know if that's alright
@rushatgabhane @cead22 @tgolen

**Comment by tgolen** [comment]
> in case this function was called by another compoenent in the future

This is a good point. I think that regardless, having it return an empty set of options would be unexpected to matter what is calling it.

**Comment by s77rt** [comment]
@tgolen If you search for nothing you should expect nothing. That seems to be the behaviour of other options list.

So should we keep the early return check or not?

**Comment by s77rt** [comment]
I think we are discussion for a future usage, as for now, returning early or not does not have an effect after applying @cead22 suggestion.
Let's get this resolved asap, should I revert my last commit (the early return) or not?

> Returning true feels wrong. The string isn't matching

This is arguable same as here `"anything".includes("")` this will return true

**Comment by s77rt** [comment]
I have to proceed with making the new videos for the linked PR. So I'm leaving the early return for now, and it can be reverted without introducing any change on the functionality

**Comment by s77rt** [comment]
@cead22 Your suggestion won't work, the early return does not return the input itself, it returns an object that includes the input.

Meaning we have to change `this.getCurrencyOptions(this.props.currencyList);` to `{currencyOptions: this.getCurrencyOptions(this.props.currencyList)};` which looks uglier and hard to maintain e.g. what if we added another value to the object? we will have to change two files 

So far, it seems better to things as they are and relay on the early return.

If you still think we should apply your suggestion please submit the suggestion on the PR (should be ready soon)

**Comment by s77rt** [comment]
[PR](https://github.com/Expensify/App/pull/13295) is ready for review

**Comment by s77rt** [comment]
a bit unrelated question
the original PR was merged within 3 days so the bonus would apply in that case.
However the PR introduced this regression, does this mean the bonus is no longer applicable?

**Comment by cead22** [comment]
@s77rt thanks!

> @cead22 Your suggestion won't work, the early return does not return the input itself, it returns an object that includes the input.
> 
> Meaning we have to change `this.getCurrencyOptions(this.props.currencyList);` to `{currencyOptions: this.getCurrencyOptions(this.props.currencyList)};` which looks uglier and hard to maintain e.g. what if we added another value to the object? we will have to change two files

@tgolen @rushatgabhane curious for your thoughts on this. I'm good with the original proposal, but I could go either way on this

- re: uglier, I think this is a personal preference, and being explicit isn't always a bad thing
- re: hard to maintai, ca you elaborate? it's not obvious to me why that's the case
- re: "what if we added another value to the object" I think is a premature optimization, but I'm curious, do we anticipate adding another value to this object?

**Comment by s77rt** [comment]
@cead22 I meant if we added another value here
https://github.com/Expensify/App/blob/d7bc6501d6ee282de24850313c00fc7b41999fcf/src/libs/OptionsListUtils.js#L793-L796

I don't see much performance gain, so I don't know if it's worth the change.. Whether to make the change or not is up to you, This won't effect the current functionality.

**Comment by tgolen** [comment]
Come to think of it, I sorta think that having `getCurrencyListForSections` inside `OptionsListUtils.js` is not really the right place for it. If it's called only once, by one component, then it should just live inside the component and it doesn't need to be in a utils file. This would solve the concern of maintenance.

I'm not concerned about adding other values to that return value because that is a preoptimization.

If the method is moved into the component, then I would simplify the return to just be:

```
return filteredOptions;
```

**Comment by cead22** [comment]
I agree with that, and we don't need to use `isSearchStringMatch` in `getCurrencyListForSections`, because we don't need to remove spaces or commas, right?

To clarify what I'm suggesting in my previous comment, it's below, and the idea is to avoid calling `OptionsListUtils.getCurrencyListForSections` passing an empty string as a second param which makes it so we return everything returned by the first param `this.getCurrencyOptions(...)`
```diff
diff --git a/src/pages/iou/IOUCurrencySelection.js b/src/pages/iou/IOUCurrencySelection.js
index 2621b19ce..7db788d44 100644
--- a/src/pages/iou/IOUCurrencySelection.js
+++ b/src/pages/iou/IOUCurrencySelection.js
@@ -41,7 +41,7 @@ class IOUCurrencySelection extends Component {
     constructor(props) {
         super(props);
 
-        const {currencyOptions} = OptionsListUtils.getCurrencyListForSections(this.getCurrencyOptions(this.props.currencyList), '');
+        const currencyData = {currencyOptions: this.getCurrencyOptions(this.props.currencyList)};
 
         this.state = {
             searchValue: '',
@@ -90,10 +90,10 @@ class IOUCurrencySelection extends Component {
      * @return {void}
      */
     changeSearchValue(searchValue) {
-        const {currencyOptions} = OptionsListUtils.getCurrencyListForSections(
-            this.getCurrencyOptions(this.props.currencyList),
-            searchValue,
-        );
+        const currencyData = searchValue
+            ? OptionsListUtils.getCurrencyListForSections(this.getCurrencyOptions(this.props.currencyList), searchValue)
+            : {currencyOptions: this.getCurrencyOptions(this.props.currencyList)};
+
         this.setState({
             searchValue,
             currencyData: currencyOptions,

```

**Comment by s77rt** [comment]
@tgolen
 - We can't move it because it relies on `isSearchStringMatch` If you think we can move it please post a code

@cead22 
- `isSearchStringMatch` is used to check for a match not spaces, commas,... it's required.
- the use of `currencyData` is better but your code is not complete

**Comment by cead22** [comment]
@s77rt sorry I don't think I explained myself correctly -- why do we need any of the following functionality from `isSearchStringMatch`?
- `.replace(/,/g, ' ')`
- `_.map( .split(' '), word => word.trim())
- `.replace(new RegExp(/&nbsp;/g), '');`
- `Str.escapeForRegExp(word)`
- `(!isChatRoom && participantNames.has(word)`

I don't think we need any of that when searching for currencies, and I'm proposing we update `getCurrencyListForSections` like this
```diff
index bc9a4facd..bcb72b084 100644
--- a/src/libs/OptionsListUtils.js
+++ b/src/libs/OptionsListUtils.js
@@ -788,7 +788,8 @@ function getHeaderMessage(hasSelectableOptions, hasUserToInvite, searchValue, ma
  */
 function getCurrencyListForSections(currencyOptions, searchValue) {
     const filteredOptions = _.filter(currencyOptions, currencyOption => (
-        isSearchStringMatch(searchValue, currencyOption.text)));
+        let regex = new RegExp(/searchValue/i);
+        return regex.test(currencyOption.text)));

     return {
         // returns filtered options i.e. options with string match if search text is entered

```

Does that make sense?

**Comment by s77rt** [comment]
@cead22 Thanks! Yeah that makes sense

**Comment by tgolen** [comment]
Great, I like the simplification. At this point, could a fully updated proposal be posted please?

**Comment by cead22** [comment]
+1

**Comment by s77rt** [comment]
@tgolen thanks for taking care of this

**Comment by chiragsalian** [comment]
This is no longer a deploy blocker since the code is on staging. Removing deploy blocker label.

**Comment by mallenexpensify** [comment]
@aimane-chnaif can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01cfd75a5dd3d584dd

**Comment by mallenexpensify** [comment]
OK, I think this can be closed then.  I'm leaving the duplicate 'hold for payment' notifications in the title so we can see why it's happening, I'll create a GH if a fix is needed. Discussing [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1672102350248879).
